endpoints/s3/src/completemultipartupload.cpp

```diff
@@ -404,27 +432,48 @@ void irods::s3::actions::handle_completemultipartupload(
 
 	// wait until all threads are complete
 	std::unique_lock<std::mutex> lk(cv_mutex);
-	cv.wait(lk, [&upload_status_object, max_part_number, func = __FUNCTION__]() {
+	cv.wait(lk, [&upload_status_object, max_part_number, func = __func__]() {
 		logging::debug("{}: wait: task_done_counter is {}", func, upload_status_object.task_done_counter);
 		return upload_status_object.task_done_counter == max_part_number;
 	});
 
-	d.close();
+	// close the object and delete the entry in the replica_token_and_number_map
+	if (part_shmem::replica_token_and_number_map.find(upload_id) != part_shmem::replica_token_and_number_map.end()) {
+		// Read all of the shared pointers in the tuple to make sure they are destructed in
+		// the order we require. std::tuple does not guarantee order of destruction.
+		auto conn_ptr = std::get<2>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto transport_ptr = std::get<3>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto dstream_ptr = std::get<4>(part_shmem::replica_token_and_number_map[upload_id]);
+
+		logging::debug("{}:{} Closing iRODS file.", __func__, __LINE__);
+		if (dstream_ptr) {
+			dstream_ptr->close();
+		}
 
-	// remove the temporary part files - should we do this on failure?
-	for (int i = 0; i < max_part_number; ++i) {
-		std::remove(part_info_vector[i].part_filename.c_str());
+		// delete the entry
+		part_shmem::replica_token_and_number_map.erase(upload_id);
 	}
 
 	// check to see if any threads failed
 	if (upload_status_object.fail_flag) {
-		logging::error("{}: {}", __FUNCTION__, upload_status_object.error_string);
+		logging::error("{}: {}", __func__, upload_status_object.error_string);
 		response.result(beast::http::status::internal_server_error);
-		logging::debug("{}: returned {}", __FUNCTION__, response.reason());
+		logging::debug("{}: returned {}", __func__, response.reason());
 		session_ptr->send(std::move(response));
 		return;
 	}
 
+	// remove the temporary part files - on failures we don't want to clean up as this could be resent
+	for (int i = 0; i < max_part_number; ++i) {
+		std::remove(part_info_vector[i].part_filename.c_str());
```

Comment:
Is this missing some parameters? I don't see any `std::remove` overloads like this... https://en.cppreference.com/w/cpp/algorithm/remove

Reply:
This is the std::remove() to remove a file from here:

https://en.cppreference.com/w/cpp/io/c/remove

Reply:
Trying to resolve this.